qcoro/qcorotask.h

```diff
@@ -614,31 +650,55 @@ class Task {
     }
 
     template<typename ThenCallback, typename ErrorCallback, typename R = cb_invoke_result_t<ThenCallback, T>>
-    auto thenImpl(ThenCallback &&thenCallback, ErrorCallback &&errorCallback) -> std::conditional_t<detail::isTask_v<R>, R, Task<R>> {
+    auto thenImpl(QObject *context, ThenCallback &&thenCallback, ErrorCallback &&errorCallback) -> std::conditional_t<detail::isTask_v<R>, R, Task<R>> {
+        bool contextDeleted = false;
+        QMetaObject::Connection contextConn;
+        if (context) {
+            // Listen for the deletion of the context object.
+            // If the context object is deleted before the awaited task finishes,
+            // don't call the callback.
+            contextConn = QObject::connect(context, &QObject::destroyed, [&] {
+                contextDeleted = true;
+            });
+            // If we don't disconnect before returning,
+            // we create another problem like the one we are trying to fix.
+            // Then the lambda above would mutate contextDeleted in a deleted stack frame.
+        }
+
+        detail::Defer cleanup([contextConn]() {
+            QObject::disconnect(contextConn);
+        });
+
         const auto thenCb = std::forward<ThenCallback>(thenCallback);
         const auto errCb = std::forward<ErrorCallback>(errorCallback);
         if constexpr (std::is_void_v<value_type>) {
             try {
                 co_await *this;
             } catch (const std::exception &e) {
-                co_return handleException<R>(errCb, e);
+                if (!contextDeleted)
+                    co_return handleException<R>(errCb, e);
             }
             if constexpr (detail::isTask_v<R>) {
-                co_return co_await invokeCb(thenCb);
+                if (!contextDeleted)
+                    co_return co_await invokeCb(thenCb);
             } else {
-                co_return invokeCb(thenCb);
+                if (!contextDeleted)
+                    co_return invokeCb(thenCb);
             }
         } else {
             std::optional<T> value;
             try {
                 value = co_await *this;
             } catch (const std::exception &e) {
-                co_return handleException<R>(errCb, e);
+                if (!contextDeleted)
+                    co_return handleException<R>(errCb, e);
             }
             if constexpr (detail::isTask_v<R>) {
-                co_return co_await invokeCb(thenCb, std::move(*value));
+                if (!contextDeleted)
+                    co_return co_await invokeCb(thenCb, std::move(*value));
             } else {
-                co_return invokeCb(thenCb, std::move(*value));
+                if (!contextDeleted)
+                    co_return invokeCb(thenCb, std::move(*value));
             }
         }
```

Comment:
Interesting that this triggers a warning in clang but not in GCC that the execution can reach an end of a non-void coroutine without `co_return`.

You could `co_return R{}`, but if course that will cause a compilation error if `R` is not default-constructible.

Reply:
After pushing this I noticed that I'm relying on something that might be a gcc bug here, so not sure if this approach makes much sense.

Reply:
There's definitely a difference between clang and gcc: https://godbolt.org/z/vf3enTv7b - both compilers are able to compile the code, but only clang emits a warning. I think it's the same thing as omitting a `return` at the end of a non-void function: the code compiles, but the function's return value will be undefined.

I'll search the GCC bugzilla tomorrow to see if it's been reported already, it's too late here now :)